### DISCUSSION TITLE
How to apply multiple filters to same exists sub query?

### DISCUSSION THREAD

**Discussion by IndexZer0** [abbasudo/laravel-purity#52] [Q&A] [answered]
See example below.

Am I crafting the filters in the correct format here?

```php
$filters = [
    'name' => [
        '$in' => [
            'Name'
        ],
    ],
    'books' => [
        'name' => [
            '$in' => [
                'Game'
            ],
        ],
        'description' => [
            '$in' => [
                'Thrones',
            ],
        ],  
    ],
],

$results = Author::filter($filters)->get();

Abbasudo\Purity\Exceptions\FieldNotSupported: The field 'books' is not supported for model App\Models\Book.
```

**Comment by IndexZer0** [comment]
@abbasudo - Would be great for your input here. There is no example in the documentation on filtering by multiple relationship properties.

When I try the above, it fails when trying to apply the second filter inside the books array.

`name` gets applied to the books exists sub query
but `description` fails.

Am I using this package wrong or is this a limitation?

**Reply by abbasudo** [reply]
Hey @IndexZer0 
your query to filter results is correct and should be working properly as you described.
provide me with the raw SQL query by adding `->dd()` to the end of the your chain insted of `get()` so i can assess you or fix any potential bugs with purity.

**Reply by IndexZer0** [reply]
@abbasudo 

```php
class Author extends Model
{
    use Filterable;
    use Sortable;

    protected $filterFields = [
        'name', // field
        'books', // relation
    ];

    public function books(): HasMany
    {
        return $this->hasMany(Book::class);
    }
}

class Book extends Model
{
    use Filterable;
    use Sortable;

    protected $filterFields = [
        'name', // field
        'description', // field
    ];

    public function author(): BelongsTo
    {
        return $this->belongsTo(Author::class);
    }
}

$filters = [
    'name' => [
        '$in' => [
            'Name'
        ],
    ],
    'books' => [
        'name' => [
            '$in' => [
                'Game'
            ],
        ],
        'description' => [
            '$in' => [
                'Thrones',
            ],
        ],
    ],
];

$results = Author::with(['books'])
    ->filter($filters)
    ->dd();
```

If I set `silent` to true in the config, and then `->dd()`. I see this:

```sql
select * from `authors` where `name` in (?) and exists (select * from `books` where `authors`.`id` = `books`.`author_id` and `name` in (?))
```

```php
array:2 [
  0 => "Name"
  1 => "Game"
]
```

It hasn't added the `description` where clause to the query and when `silent` is set to false, package throws `FieldNotSupported` exception

**Reply by abbasudo** [reply]
@IndexZer0 
change input to this and send me the results 
```php
$filters = [
    'name' => [
        '$in' => [
            'Name'
        ],
    ],
    'books' => [
        'name' => [
            '$in' => [
                'Game'
            ],
        ],
    ],
    'books' => [
        'description' => [
            '$in' => [
                'Thrones',
            ],
        ],
    ],
];

```

**Reply by abbasudo** [reply]
@num8er any thoughts?

**Reply by num8er** [reply]
@IndexZer0  I cannot reproduce issue:
![image](https://github.com/abbasudo/laravel-purity/assets/603434/9ee72da5-4a9b-41b6-b748-a15cf6b0dd83)

See PR: https://github.com/abbasudo/laravel-purity/pull/55

**Reply by abbasudo** [reply]
@IndexZer0 are you running the latest version of Purity?

**Reply by IndexZer0** [reply]
Stack:

- PHP 8.2.17
- Laravel v10.48.7
- Purity v3.2.0

**Reply by IndexZer0** [reply]
@num8er 

That is strange that you're not getting a `FieldNotSupported` exception, yet I am.

I see you're running php 8.3, which version of laravel for executing that test?

My stack where the issue is happening: https://github.com/abbasudo/laravel-purity/discussions/52#discussioncomment-9175440

---

I see that purity is supporting many version of laravel.

`"laravel/framework": "^11.0|^10.0|^9.0|^8.0"`

Maybe worth making the test github action execute on various configurations? 

https://github.com/spatie/package-skeleton-laravel/blob/main/.github/workflows/run-tests.yml

---

However, the query that is generated seems counter intuative to me.

Why two different exist sub queries?

You have `books` defined once in the `$filters` array.

I would expect a single `and exists (select * from books)`, and inside that subquery it should be checking `name` and `description`.

**Reply by num8er** [reply]
I'm running unit test inside the package, so based on composer.json it's latest - Laravel 11.x

I agree need to check with different versions of Laravel, but the thing is whoever will use this package - will use latest Laravel (10, 11)

**Reply by num8er** [reply]
@IndexZer0 btw my test is merged to master branch.
Can You just clone the repo and run `./vendor/bin/phpunit` ?

**Reply by IndexZer0** [reply]
I did just fork the repo and run the test and is working for me inside the package.

Throwing `FieldNotSupported` exception when using package inside a laravel app.

Odd one.

**Reply by IndexZer0** [reply]
@num8er - Any thoughts on what I mentioned about the way the query is constructed with two exists sub queries?

**Reply by num8er** [reply]
> @num8er - Any thoughts on what I mentioned about the way the query is constructed with two exists sub queries?

Cause there is no $or filter, so it made 2 different queries.

**Reply by num8er** [reply]
> I did just fork the repo and run the test and is working for me inside the package.
> 
> Throwing `FieldNotSupported` exception when using package inside a laravel app.
> 
> Odd one.

Make sure Your models and etc not cached, maybe worth to try clean artisan caches.

**Reply by IndexZer0** [reply]
Not trying to do an or query. Doesn't it make sense that both the name and description clauses should happen within the same exists sub query?

Why do exists twice?

**Comment by IndexZer0** [comment]
> @IndexZer0 change input to this and send me the results
> 
> ```
> $filters = [
>     'name' => [
>         '$in' => [
>             'Name'
>         ],
>     ],
>     'books' => [
>         'name' => [
>             '$in' => [
>                 'Game'
>             ],
>         ],
>     ],
>     'books' => [
>         'description' => [
>             '$in' => [
>                 'Thrones',
>             ],
>         ],
>     ],
> ];
> ```

You've got `books` as an array key twice so invalid array. Anyway, here is what happens.

```sql
"select * from `authors` where `name` in (?) and exists (select * from `books` where `authors`.`id` = `books`.`author_id` and `description` in (?))"
```

```php
array:2 [
  0 => "Name"
  1 => "Thrones"
]
```

**Reply by num8er** [reply]
@IndexZer0  You're using same key `books` twice in assoc array. So it's not possible and last one will replace it.

Your initial array is correct:
```php
$filters = [
    'name' => [
        '$in' => [
            'Name'
        ],
    ],
    'books' => [
        'name' => [
            '$in' => [
                'Game'
            ],
        ],
        'description' => [
            '$in' => [
                'Thrones',
            ],
        ],  
    ],
],
```

**Reply by IndexZer0** [reply]
Lol yes I know. @abbasudo told me to to run that example and I pointed that out. 🙂

**Reply by num8er** [reply]
@IndexZer0  @abbasudo  I made another test where I'm requesting using http and passing filters via query string.
So both of them generate same sql queries:

![image](https://github.com/abbasudo/laravel-purity/assets/603434/d2b6ecf9-00e0-402f-b2cf-a0cc55a0feea)

**Comment by IndexZer0** [comment, marked as answer]
@abbasudo 

Okay, had a screenshare with @num8er 

We found that the bug does not exist in dev-master when i composer require in my application.

v3.2.0 is broken.

v3.2.0 was released containing bug. PR #49 fixed issue. no release done since.

Can a release be done from current master please.

**Reply by num8er** [reply]
@abbasudo I propose tests to written in non silent mode. 
Cause since silent mode hides most of issues like with @IndexZer0  we are having unexpected, invisible bugs.

**Reply by IndexZer0** [reply]
@num8er Thanks for working with me to find cause of issue.

**Reply by num8er** [reply]
@IndexZer0 welcome :)

**Reply by abbasudo** [reply]
@IndexZer0 that's great to know. thanks for informing me. just drafted a new release.

**Comment by abbasudo** [comment]
> @abbasudo I propose tests to written in non silent mode. Cause since silent mode hides most of issues like with @IndexZer0 we are having unexpected, invisible bugs.

@num8er Good idea. thanks for your help.
